### DISCUSSION TITLE
ScriptAPI/Endstone plugins bridge support

### DISCUSSION THREAD

**Discussion by leaftail1880** [EndstoneMC/endstone#10] [Ideas]
BDS has support for [Scripting API](https://learn.microsoft.com/en-us/minecraft/creator/documents/scriptingservers?view=minecraft-bedrock-stable) (JS/TS) that comes with the behavior packs since the 1.19 and something like a bridge between the endstone and the ScriptAPI would be very nice for the developers. 

From the bdsx (other bds mod) discord server:

> The benefit of this over other addon communication system is that it can be done synchronously
>
> We were able to register a function globally so that it doesn't have to be imported, however in a more realistic setting you would be registering a module and adding functions to it
> 
> The addon side of my code looked like this:
> ```js
> import { system } from "@minecraft/server";
>
> system.afterEvents.scriptEventReceive.subscribe((event) => {
>    if (event.id !== "test") {
>        return;
>    }
>
>    let res = nativeFunction();
>
>    console.log(`Ran and returned: ${res}`);
> });
> ```
> And the C++ function that's being called looks like this: 
> ```cpp
> extern "C" {
>    JSValue nativeFunction(JSContext *ctx, JSValue this_val, int argc, JSValue *argv) {
>        std::cout << "I'm a custom C++ function being called from an addon!\n";
>
>        return fromUInt(10);
>    }
> }
>```

**Comment by wu-vincent** [comment]
Thank you for the great suggestion. It would surely be an exciting feature if endstone is able to interact with the scripting api system. This would definitely benefit developers who would like to extend the scripting api with endstone plugins or the other way around.

With that said, I don't have much experience with javascript. I will probably look into the scripting system eventually but in the meantime, please feel free to contribute if you have any progress on this. I would really appreciate it!

**Reply by leaftail1880** [reply]
More context & code sources on bdsx discord server: https://discord.com/channels/646456965983240212/1204828139159691315/1205196342297890826
